syntax = "proto3";

package kvstore.server;

import "google/protobuf/timestamp.proto";

message Introduction {
  string ip = 1;
  uint32 port = 2;
}

message DoWrite {
  string uuid = 1;
  uint32 key = 2;
  google.protobuf.Timestamp timestamp = 3;
  string val = 4;
}

message WriteComplete {
  string uuid = 1;
}

message DoRead {
  string uuid = 1;
  uint32 key = 2;
}

message ReadComplete {
  string uuid = 1;
  google.protobuf.Timestamp timestamp = 2;
  string val = 3;
}

message ReadRepair {
  uint32 key = 1;
}

message ReadRepairResponse {
  uint32 key = 1;
  google.protobuf.Timestamp timestamp = 2;
  string val = 3;
}

message ServerMessage {
  oneof msg {
    Introduction intro = 1;
    DoWrite do_write = 2;
    WriteComplete write_complete = 3;
    DoRead do_read = 4;
    ReadComplete read_complete = 5;
    ReadRepair read_repair = 6;
    ReadRepairResponse read_repair_response = 7;
  }
}
